### PR TITLE
OpticalFlow: PX4FLOW Linux support

### PR DESCRIPTION
**Pull request by vmayoral**
Driver coded for Linux and tests updated.

As discussed in https://github.com/mavlink/mavros/issues/271, the driver seems to go in the right direction but it still does not refresh the data as expected.

### CODE REVIEW

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      hal.scheduler->panic(PSTR("PX4FLOW: unable to get semaphore"));
+  }
+
+  // to be sure this is not a ll40ls Lidar (which can also be on
+  // 0x42) we check if a I2C_FRAME_SIZE byte transfer works from address
+  // 0. The ll40ls gives an error for that, whereas the flow
+  // happily returns some data
+  uint8_t val[I2C_FRAME_SIZE];
+  if (hal.i2c->readRegisters(I2C_FLOW_ADDRESS, 0, I2C_FRAME_SIZE, val))
+    hal.scheduler->panic(PSTR("ll40ls Lidar"));
```
Comment (tridge) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: we should not be using panic() here. Instead set a flag saying it hasn't found the sensor and then fail the read()
All of the panic() calls should be removed from this code

**Comment by tridge** [comment]
just a general comment on panic(). I've noticed you using panic() as an error handler in other PRs as well. panic() should only be used for fatal programming errors where the vehicle definitely can't fly. It should not be used for optional sensors.

**Comment by vmayoral** [comment]
Thanks @tridge. Will put some more time into it this weekend.

**Comment by vmayoral** [comment]
Fixed @tridge's comments and tried a few things changing the reading method. No progress so far, incoming data is always the same. 

@LorenzMeier, mind taking a look?

**Comment by vmayoral** [comment]
@tridge, @magicrub finally found the issues. Code works now with AP_HAL_Linux.
Let me know if you'd like any change on it.

Regards,

**Comment by rmackay9** [comment]
@vmayoral,
I see your last comment is on May 17th but the last commit shown is from a month earlier (Apr 19th).  Does the commit in this PR reflect your latest work?  I think perhaps not because it includes issues that Tridge has pointed out are incorrect (i.e. using PANIC in the code).

**Comment by rmackay9** [comment]
I'm closing this because I've raised a new PR that includes your work and some work I've done on top of it.   https://github.com/diydrones/ardupilot/pull/2667

**Pull request closed without merging**